README.md
# Sinatra ActiveRecord Extension

Extends [Sinatra](http://www.sinatrarb.com/) with extension methods and Rake
tasks for dealing with an SQL database using the
[ActiveRecord ORM](https://github.com/rails/rails/tree/master/activerecord).

![test badge](https://github.com/sinatra-activerecord/sinatra-activerecord/workflows/rspec/badge.svg)

## Requirement
ActiveRecord >= 4.1

## Setup

Put it in your `Gemfile`, along with the adapter of your database. For
simplicity, let's assume you're using SQLite:

```ruby
gem "sinatra-activerecord"
gem "sqlite3"
gem "rake"
```

Now require it in your Sinatra application, and establish the database
connection:

```ruby
# app.rb
require "sinatra/activerecord"

set :database, {adapter: "sqlite3", database: "foo.sqlite3"}
# or set :database_file, "path/to/database.yml"
```

If you have a `config/database.yml`, it will automatically be loaded, no need
to specify it. Also, in production, the `$DATABASE_URL` environment variable
will automatically be read as the database (if you haven't specified otherwise).

If both `config/database.yml` and `$DATABASE_URL` are present, the database configuration of this two will be merged, with $DATABASE_URL's variable taking precedence over database.yml (for the same variable / key).


Note: If you are using ActiveRecord 6.0 and above, and have [defined multiple databases](https://guides.rubyonrails.org/active_record_multiple_databases.html#setting-up-your-application) for the database.yml, the $DATABASE_URL configuration will be discarded, following [Active Record convention here](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/database_configurations.rb#L169).


Note that in **modular** Sinatra applications you will need to first register
the extension:

```ruby
class YourApplication < Sinatra::Base
  register Sinatra::ActiveRecordExtension
end
```

Now require the rake tasks and your app in your `Rakefile`:

```ruby
# Rakefile
require "sinatra/activerecord/rake"

namespace :db do
  task :load_config do
    require "./app"
  end
end
```

In the Terminal test that it works:

```sh
$ bundle exec rake -T
rake db:create            # Create the database from DATABASE_URL or config/database.yml for the current Rails.env (use db:create:all to create all dbs in the config)
rake db:create_migration  # Create a migration (parameters: NAME, VERSION)
rake db:drop              # Drops the database using DATABASE_URL or the current Rails.env (use db:drop:all to drop all databases)
rake db:fixtures:load     # Load fixtures into the current environment's database
rake db:migrate           # Migrate the database (options: VERSION=x, VERBOSE=false)
rake db:migrate:status    # Display status of migrations
rake db:rollback          # Rolls the schema back to the previous version (specify steps w/ STEP=n)
rake db:schema:dump       # Create a db/schema.rb file that can be portably used against any DB supported by AR
rake db:schema:load       # Load a schema.rb file into the database
rake db:seed              # Load the seed data from db/seeds.rb
rake db:setup             # Create the database, load the schema, and initialize with the seed data (use db:reset to also drop the db first)
rake db:structure:dump    # Dump the database structure to db/structure.sql
rake db:version           # Retrieves the current schema version number
```

And that's it, you're all set :)

## Usage

You can create a migration:

```sh
$ bundle exec rake db:create_migration NAME=create_users
```

This will create a migration file in your migrations directory (`./db/migrate`
by default), ready for editing.

```ruby
class CreateUsers < ActiveRecord::Migration
  def change
    create_table :users do |t|
      t.string :name
    end
  end
end
```

Now migrate the database:

```sh
$ bundle exec rake db:migrate
```

You can also write models:

```ruby
class User < ActiveRecord::Base
  validates_presence_of :name
end
```

You can put your models anywhere you want, only remember to require them if
they're in a separate file, and that they're loaded after `require "sinatra/activerecord"`.

Now everything just works:

```ruby
get '/users' do
  @users = User.all
  erb :index
end

get '/users/:id' do
  @user = User.find(params[:id])
  erb :show
end
```

A nice thing is that the `ActiveRecord::Base` class is available to
you through the `database` variable:

```ruby
if database.table_exists?('users')
  # Do stuff
else
  raise "The table 'users' doesn't exist."
end
```

## History

This gem was made in 2009 by Blake Mizerany, creator of Sinatra.

## Social

You can follow Janko on Twitter, [@jankomarohnic](http://twitter.com/jankomarohnic).
You can follow Axel on Twitter, [@soulchildpls](http://twitter.com/soulchildpls).

## License

[MIT](https://github.com/sinatra-activerecord/sinatra-activerecord/blob/master/LICENSE)


CHANGELOG.md
# Changelog

## Version 2.0.26
Removed unused variables, and removed support for ruby 2.5

## Version 2.0.25
- Added support for ActiveRecord 7.0

## Version 2.0.24
- Fixed compatibility with Psych version 4.0.0+ , use unsafe_load if available.
- Added Ruby 3.0 and head (latest version of Ruby) to the CI matrix
Thanks Onaka (@onk) for contributing to this.

## Version 2.0.23
- For ActiveRecord 6.0 and above, if the the configuration (database.yml) has multiple databases defined, it will discard the DATABASE_URL spec (if it is supplied from environment variables), following [Active Record convention here](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/database_configurations.rb#L169). Thanks JoakimKlaxit (@JoakimKlaxit) for contributing to this.


## Version 2.0.22

- Added compability fix for ActiveRecord 6.1 ([#107](https://github.com/janko-m/sinatra-activerecord/issue/107)), as they have moved ConnectionUrlResolver to another module. Thanks Richard Peck(@richpeck) for contributing to this.

## Version 2.0.21

- If both config/database.yml and $DATABASE_URL is present, the database configuration will be merged from this two, and $DATABASE_URL's variables will take precedence. (https://github.com/sinatra-activerecord/sinatra-activerecord/pull/103)

## Version 2.0.9

- Don't try to load config/database.yml if `$DATABASE_URL` is present ([#55](https://github.com/janko-m/sinatra-activerecord/pull/55), thanks to @exviva)

## Version 2.0.8

- Use the ConnectionManagement AR middelware for clearing active connections

## Version 2.0.7

- Allow multiple databases to be used ([#54](https://github.com/janko-m/sinatra-activerecord/pull/54))

## Version 2.0.6

* Prefer `DATABASE_URL` to config/database.yml ([#47](https://github.com/janko-m/sinatra-activerecord/issue/47)).

## Version 2.0.5

* Disabled ActiveRecord logging in tests

## Version 2.0.4

* You're now encouraged to load the app in the `:load_config` Rake task. This
  way your app will only be required when executing ActiveRecord tasks, and
  not other potential tasks which don't need the database. Your Rakefile can
  now be updated to this:

  ```rb
  require "sinatra/activerecord/rake"

  namespace :db do
    task :load_config do
      require "./app" # or whatever your app name is
    end
  end
  ```

* ActiveRecord isn't logged in production anymore.

## Version 2.0.3

- Use `File.exist?` instead of `File.exists?` to avoid warnings (thanks to [**@matthiase**](https://github.com/matthiase)).

## Version 2.0.2

- Allow setting the environment as a String, aside from Symbol.

## Version 2.0.1

- Fixed the issue where it mattered in which order `sinatra/activerecord/rake`
  and `./app` are required.

## Version 2.0.0

- ActiveRecord 4.1 is now supported.
- Rake tasks are now directly imported from ActiveRecord, which means *all* of
  the tasks are now available.

## Version 1.7.0

- Added `db:reset` and `db:migrate:reset` tasks ([#27](https://github.com/janko-m/sinatra-activerecord/pull/27)).
  Thanks to @csgavino.

## Version 1.6.0

- Added ability to specify folder for the database ([#26](https://github.com/janko-m/sinatra-activerecord/pull/26)).
  Thanks to @jamesrwhite.

## Version 1.5.0

- Added ability to specify `:sql` format of the schema, as well as
  `db:structure` tasks ([#25](https://github.com/janko-m/sinatra-activerecord/pull/25)).
  Thanks to @icambron.

## Version 1.4.0

- Added `db:test:prepare` Rake task.

## Version 1.3.0

- Added `db:create`, `db:drop` and `db:reset` Rake tasks.

## Version 1.2.5

- Fix a bug with ActiveRecord 3.0.

## Version 1.2.4

- By default look for a `config/database.yml` in the root of the running process.

## Version 1.2.3

- Allow ActiveRecord 4

## Version 1.2.2

- Fix not allowing environments other than "development", "test" and "production".

## Version 1.2.1

- Fix specifying database from a file not working properly.

## Version 1.2.0

- You can now set a custom migrations directory. Take a look at
[this wiki](https://github.com/janko-m/sinatra-activerecord/wiki/Changing-the-migrations-directory).

- You can now set the database to a YAML file. Take a look at
[this wiki](https://github.com/janko-m/sinatra-activerecord/wiki/Alternative-database-setup).

## Version 1.1.2

- Users are now warned if they use SQLite URLs with 2 slashes.

## Version 1.1.1

- The `set :database` command now accepts everything that
  `ActiveRecord::Base.establish_connection` accepts. A Hash
  is one example:

```ruby
set :database, {
  adapter: "sqlite3",
  database: "db/foo.db",
  pool: 5,
  timeout: 5000
}
```

- The database is now reset whenever you specify it with `set :database`.
- The migrations are now logged just like in Rails.

## Version 1.1.0

- If no database is specified, it will try to read from the `DATABASE_URL`
  environment variable (which is often set in production). This was
  removed in 1.0.0, but now I'm bringing it back.

## Version 1.0.1

- Removed deprecation warnings when using the `#database` helper.

## Version 1.0.0

- Fixed the database not working. Sorry, I won't let it happen again.

### Changes that are NOT backwards compatible

- There is no more "default" database. Now you always have to specify
  the database you want to connect to, it doesn't default to
  `"sqlite://#{environment}.db"` anymore.

- When you're setting an SQLite database, you now have to put 3 slashes
  after `sqlite:` instead of 2. So instead of `sqlite://database.db` you
  have to write `sqlite:///database.db`. This is now a valid URL.

## Version 0.2.1 (yanked)

- The previous version was yanked, because I forgot to add the
  `db:rollback` task (I didn't figure out how to test it yet, otherwise
  I would know it was missing). This version is then actually 0.2.0.

## Version 0.2.0 (yanked)

- Added `db:rollback` rake task.

- If you're using SQLite, you can now specify the path to where
  you want your database file to be (thanks to **@mpalmer** for this). Refer to

spec/spec_helper.rb
begin
  require "pry"
rescue LoadError
end

ROOT = File.expand_path("../", File.dirname(__FILE__))
Dir[File.join(ROOT, "spec/support/**/*.rb")].each &method(:require)

RSpec.configure do |config|

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end

  config.before(:each) do
    FileUtils.mkdir_p("tmp")
    FileUtils.rm("tmp/foo.sqlite3") if File.exists?("tmp/foo.sqlite3")
  end

  config.after(:each) do
    FileUtils.rm_rf("tmp")
  end

  config.include Helpers
end


spec/sinatra/activerecord_spec.rb
require 'spec_helper'
require 'sinatra/base'
require 'fileutils'

RSpec.describe "the sinatra extension" do

  let(:url_sqlite_db_name) { 'foo_baz' }
  let(:database_url) do
    if ActiveRecord::VERSION::STRING.start_with? "4.1"
      "sqlite3:tmp/#{url_sqlite_db_name}.sqlite3"
    else
      "sqlite3:///tmp/#{url_sqlite_db_name}.sqlite3"
    end
  end

  let(:app) do
    Class.new(Sinatra::Base) do
      set :root, nil
      register Sinatra::ActiveRecordExtension
    end
  end

  before do
    hide_const("Rake")
    ActiveRecord::Base.remove_connection
  end

  after do
    ActiveRecord::Base.logger = nil
  end

  it "exposes ActiveRecord::Base" do
    expect(app.database).to eq ActiveRecord::Base
  end

  it "establishes the connection with a database url" do
    app.database = database_url

    expect{ActiveRecord::Base.connection}.not_to raise_error
  end

  it "establishes the connection with a hash" do
    app.database = {adapter: "sqlite3", database: "tmp/foo.sqlite3"}

    expect{ActiveRecord::Base.connection}.not_to raise_error
  end

  it "handles namespacing into environments" do
    app.environment = :development
    app.database = {development: {adapter: "sqlite3", database: "tmp/foo.sqlite3"}}

    expect{ActiveRecord::Base.connection}.not_to raise_error
  end

  it "allows settings environments as Strings" do
    app.environment = "development"
    app.database = {development: {adapter: "sqlite3", database: "tmp/foo.sqlite3"}}

    expect{ActiveRecord::Base.connection}.not_to raise_error
  end

  it "raises an appropriate error when the database spec is invalid" do
    expect{app.database = {}}.to raise_error(ActiveRecord::AdapterNotSpecified)
  end

  it "doesn't try to establish connection when database isn't set" do
    expect{app.database}.not_to raise_error
  end

  it "establishes connection from DATABASE_URL if it's present" do
    ENV["DATABASE_URL"] = database_url
    app

    expect{ActiveRecord::Base.connection}.not_to raise_error

    ENV.delete("DATABASE_URL")
  end

  it "allows specifying database through a file" do
    app.database_file = "spec/fixtures/database.yml"

    expect{ActiveRecord::Base.connection}.not_to raise_error
  end

  it "allows specifying multiple databases" do
    app.root = "spec/fixtures"
    app.database_file = "database.yml"
    app.database = database_url

    expect{ActiveRecord::Base.establish_connection(:test)}.not_to raise_error
    expect{ActiveRecord::Base.establish_connection(:another_test)}.not_to raise_error
  end

  it "allows specifying multiple databases, with URL having precedence" do
    ENV["DATABASE_URL"] = database_url
    FileUtils.mkdir_p("config")
    FileUtils.copy_file("#{Dir.pwd}/spec/fixtures/database.yml", "#{Dir.pwd}/config/database.yml")
    
    app

    # it uses the database name in the URL, instead of the database.yml
    expect(app.database.connection.raw_connection.filename).to include(url_sqlite_db_name)
    # it uses the pool size specified in yml file, as the URL didn't override it
    expect(app.database.connection_pool.size).to eq(3)

    expect{ActiveRecord::Base.connection}.not_to raise_error

    ENV.delete("DATABASE_URL")
    FileUtils.rm_rf("config")
  end

  it "allows specifying multiple environment databases, without URL precedence", if: ActiveRecord::VERSION::MAJOR >= 6 do
    ENV["DATABASE_URL"] = database_url
    FileUtils.mkdir_p("config")
    FileUtils.copy_file("#{Dir.pwd}/spec/fixtures/database.yml", "#{Dir.pwd}/config/database.yml")

    app
    ActiveRecord::Base.establish_connection(:multiple_db_test)

    # it does not use the database name in the URL, but the one defined in database.yml
    expect(app.database.connection.raw_connection.filename).to_not include(url_sqlite_db_name)

    expect{ActiveRecord::Base.connection}.not_to raise_error

    ENV.delete("DATABASE_URL")
    FileUtils.rm_rf("config")
  end

  it "expands database file path from the app root if present" do
    app.root = "spec/fixtures"
    app.database_file = "database.yml"

    expect{ActiveRecord::Base.connection}.not_to raise_error
  end

  it "doesn't expand the database file path from the app root if the path is absolute" do
    app.root = "spec/fixtures"
    app.database_file = "#{Dir.pwd}/spec/fixtures/database.yml"

    expect{ActiveRecord::Base.connection}.not_to raise_error
  end

  it "raises an error on invalid database.yml" do
    FileUtils.touch("tmp/database.yml")

    expect{app.database_file = "tmp/database.yml"}.to raise_error(ActiveRecord::AdapterNotSpecified)
  end

  it "raises an error on missing database.yml" do
    expect{app.database_file = "foo.yml"}.to raise_error(Errno::ENOENT)
  end

end


spec/sinatra/activerecord/rake_spec.rb
require 'spec_helper'
require 'fileutils'

module MigrationWithQuiet
  def migrate(*args)
    suppress_messages do
      super(*args)
    end
  end
end

ActiveRecord::Migration.send(:prepend, MigrationWithQuiet)

RSpec.describe "the rake tasks" do
  before do
    Class.new(Sinatra::Base) do
      register Sinatra::ActiveRecordExtension
      set :database, {adapter: "sqlite3", database: "tmp/foo.sqlite3"}
    end

    FileUtils.mkdir_p "db"
    FileUtils.touch "db/seeds.rb"

    require 'rake'
    require 'sinatra/activerecord/rake'

    module TaskWithReenable
      def invoke(*args)
        super(*args)
        reenable
      end
    end

    Rake::Task.send(:prepend, TaskWithReenable)
  end

  after do
    FileUtils.rm_rf "db"
  end

  it "has all the rake tasks working" do
    begin
      ENV["NAME"] = "create_users"

      Rake::Task["db:create"].invoke
      Rake::Task["db:create_migration"].invoke
      Rake::Task["db:migrate"].invoke
      Rake::Task["db:migrate:redo"].invoke
      Rake::Task["db:reset"].invoke
      Rake::Task["db:seed"].invoke

      ENV.delete("NAME")
    rescue SystemExit
      fail 'should not exit'
    end

    # ensure the migration file is created
    migration_file = Dir["#{FileUtils.pwd}/db/migrate/*_create_users.rb"]
    expect(migration_file).not_to be_empty

    schema_file = Dir["#{FileUtils.pwd}/db/schema.rb"]
    expect(schema_file).not_to be_empty

    seeds_file = Dir["#{FileUtils.pwd}/db/seeds.rb"]
    expect(seeds_file).not_to be_empty
  end

  it 'works with providing argument instead of named env param' do
    begin
      ARGV[1] = 'create_users'

      Rake::Task["db:create"].invoke
      Rake::Task["db:create_migration"].invoke
      Rake::Task["db:migrate"].invoke
      Rake::Task["db:migrate:redo"].invoke
      Rake::Task["db:reset"].invoke
      Rake::Task["db:seed"].invoke
    rescue SystemExit
      fail 'should not exit'
    end

    # ensure the migration file is created
    migration_file = Dir["#{FileUtils.pwd}/db/migrate/*_create_users.rb"]
    expect(migration_file).not_to be_empty

    schema_file = Dir["#{FileUtils.pwd}/db/schema.rb"]
    expect(schema_file).not_to be_empty

    seeds_file = Dir["#{FileUtils.pwd}/db/seeds.rb"]
    expect(seeds_file).not_to be_empty
  end
end


lib/sinatra/activerecord/rake.rb
require 'active_record'
require 'active_support/core_ext/string/strip'
require 'fileutils'

module Sinatra
  module ActiveRecordTasks
    extend self

    def create
      silence_activerecord do
        ActiveRecord::Tasks::DatabaseTasks.create(config)
      end
    end

    def drop
      silence_activerecord do
        ActiveRecord::Tasks::DatabaseTasks.drop(config)
      end
    end

    def seed
      silence_activerecord do
        load("#{db_dir}/seeds.rb")
      end
    end

    def setup
      silence_activerecord do
        create
        load_schema
        seed
      end
    end

    def create_migration(migration_name, version = nil)
      raise "No NAME specified. Example usage: `rake db:create_migration NAME=create_users`" if migration_name.nil?

      migration_number = version || Time.now.utc.strftime("%Y%m%d%H%M%S")
      migration_file = File.join(migrations_dir, "#{migration_number}_#{migration_name}.rb")
      migration_class = migration_name.split("_").map(&:capitalize).join

      FileUtils.mkdir_p(migrations_dir)
      File.open(migration_file, 'w') do |file|
        file.write <<-MIGRATION.strip_heredoc
          class #{migration_class} < ActiveRecord::Migration
            def change
            end
          end
        MIGRATION
      end
    end

    def migrate(version = nil)
      silence_activerecord do
        migration_version = version ? version.to_i : version
        ActiveRecord::Migrator.migrate(migrations_dir, migration_version)
      end
    end

    def rollback(step = nil)
      silence_activerecord do
        migration_step = step ? step.to_i : 1
        ActiveRecord::Migrator.rollback(migrations_dir, migration_step)
      end
    end

    def dump_schema(file_name = "#{db_dir}/schema.rb")
      silence_activerecord do
        ActiveRecord::Migration.suppress_messages do
          # Create file
          out = File.new(file_name, 'w')

          # Load schema
          ActiveRecord::SchemaDumper.dump(ActiveRecord::Base.connection, out)

          out.close
        end
      end
    end

    def dump_structure(file_name = "#{db_dir}/structure.sql")
      ActiveRecord::Tasks::DatabaseTasks.structure_dump(config, file_name)
    end

    def purge
      if config
        ActiveRecord::Tasks::DatabaseTasks.purge(config)
      else
        raise ActiveRecord::ConnectionNotEstablished
      end
    end

    def load_schema(file_name = "#{db_dir}/schema.rb")
      load(file_name)
    end

    def load_structure(file_name = "#{db_dir}/structure.sql")
      ActiveRecord::Tasks::DatabaseTasks.structure_load(config, file_name)
    end

    def with_config_environment(environment, &block)
      previous_environment = config_environment
      begin
        config_environment(environment)
        yield
      ensure
        config_environment(previous_environment)
      end
    end

    def db_dir
      return @db_dir if @db_dir
      self.db_dir = 'db'
    end

    def db_dir=(dir)
      @db_dir = dir
      ActiveRecord::Tasks::DatabaseTasks.db_dir = @db_dir
    end

    private

    def config_environment(env = nil)
      if env
        @config_environment = env
      else
        @config_environment ||= Sinatra::Application.environment.to_s
      end
    end

    def config
      ActiveRecord::Base.configurations[config_environment] ||
        #active record expects the keys to be strings, but connection_config returns them as symbols
        Hash[ActiveRecord::Base.connection_config.map{|key, value| [key.to_s, value]}]
    end

    def migrations_dir
      ActiveRecord::Migrator.migrations_path
    end

    def silence_activerecord(&block)
      old_logger = ActiveRecord::Base.logger
      ActiveRecord::Base.logger = nil
      yield if block_given?
      ActiveRecord::Base.logger = old_logger
    end
  end
end

load 'sinatra/activerecord/tasks.rake'


lib/sinatra/activerecord/rake/helpers.rb
$default_primary_key_type = :integer

# ==[ Schema dump ]==

# NOTE: It would be great if the following worked, but I get this error:
#       "TypeError: superclass mismatch for class SchemaDumper"
#
# class ActiveRecord::ConnectionAdapters::SchemaDumper
#   private
#     def default_primary_key?(column)
#       schema_type(column) == $default_primary_key_type
#     end
# end

class ActiveRecord::SchemaDumper

  alias_method :header_real, :header
  def header(stream)
    buffer = StringIO.new
    header_real(buffer)
    string = buffer.string

    # adjust header to keep it simple
    string = $' if string =~ /^(?=\w)/ # skip past comments
    string.sub!(/(?<=Schema)\[[^\]]*\](?=\.define)/, '') # remove version
    string << "\n"

    stream.print string
  end

  alias_method :table_real, :table
  def table(table, stream)
    buffer = StringIO.new
    table_real(table, buffer)
    string = buffer.string

    # adjust table descriptions using shorthand notation
    string.sub!(/, id: :#{$default_primary_key_type}\b/o, '') # skip primary keys with default type #!# TODO: remove when we fix the class issue above
    string.sub!(/(?:, (?:charset|collation|options): "[^"]*")*, force: :cascade/, '') # skip options
    string.gsub!(/^(.+?)("(?=, ))(.*), null: false/, '\1!\2\3') # add "!" to column name for not null
    string.gsub!(/^( *t\.)(text)( .*?)(, limit: )(\d+)/) do # adjust text types
      case $5
        when "255"        then [$1, 'tinytext'  , $3].join
        when "65535"      then [$1, 'text'      , $3].join
        when "16777215"   then [$1, 'mediumtext', $3].join
        when "4294967295" then [$1, 'longtext'  , $3].join
        else $~.join
      end
    end
    string.gsub!(/^( *t\..+?), limit: (\S+)/, '\1, \2') # "limit: x" is "x"
    string.gsub!(/^( *t\..+?), default: (.*?)(?= *$|, )/, '\1, [\2]') # "default: x" is "[x]"
    string.gsub!(/^( *t\.boolean +"\w+!"), \[false\]/, '\1') # booleans default to false, so skip
    string.gsub!(/, precision: nil/, '') # skip useless precision values
    string.gsub!(/, precision: (\d+), scale: (\d+)/, ', [\1, \2]') and # "decimal(p,s)" is "[p, s]"
    string.gsub!(/, \["(\d+)\.0"\]/, ', [\1]') # adjusts defaults like ["4.0"] to [4]
    string.gsub!(/t\.decimal ("[^"]+"), \[9, 2\], \[0\]$/, 't.digits \1') # use t.digits shortcut

    # adjust indexes
    string.sub!(/^(?= *t.index)/, "\n") # add a spacer line before indexes
    string.gsub!(/^( *t.index .*?), name: "index_[^"]+"/, '\1') # strip generated index names
    string.gsub!(/^( *t.index)( .*?), unique: true/, '\1!\2') # add "!" for unique indexes

    # line up column names
    wide =[string.scan(/^ *t\.\S+/).map(&:size).max, 14].max
    string.gsub!(/^( *(?:t\.\S+|create_table))/) {$1.ljust(wide)}

    # line up column options
    wide = string.scan(/^ *t\.(?!index)[^,\n]+/).map(&:size).max
    string.gsub!(/^( *t\.(?!index)[^,\n]+(?=,))/) { $1.ljust(wide)}

    # symbolize tables, fields, and indexes
    string.gsub!(/^( *(?:t\.\S+|create_table) +)"([^"]+)"/, '\1:\2')
    string.gsub!(/^( *t\.index!? +?) \[([^\]]+)\]/) do
      list = $2.delete('"').split(', ').map(&:to_sym)
      [$1, (solo = list.size == 1) ? ' ' : '', (solo ? list[0] : list).inspect].join
    end

    stream.print string
  end

  alias_method :foreign_keys_real, :foreign_keys
  def foreign_keys(table, stream)
    buffer = StringIO.new
    foreign_keys_real(table, buffer)
    string = buffer.string

    # adjust foreign key descriptions
    string.gsub!(/^( *add_foreign_key .*?), name: "[^"]+"/, '\1') # suppress key name
    string.gsub!(/"([^"]+)"/, ':\1') # symbolize foreign keys

    stream.print string
  end
end

# ==[ Schema load ]==



lib/sinatra/activerecord/rake/activerecord_5.rb
require_relative "helpers"

seed_loader = Class.new do
  def load_seed
    load "#{ActiveRecord::Tasks::DatabaseTasks.db_dir}/seeds.rb"
  end
end

ActiveRecord::Tasks::DatabaseTasks.tap do |config|
  config.root                   = Rake.application.original_dir
  config.env                    = ENV["RACK_ENV"] || "development"
  config.db_dir                 = "db"
  config.migrations_paths       = ["db/migrate"]
  config.fixtures_path          = "test/fixtures"
  config.seed_loader            = seed_loader.new
  config.database_configuration = ActiveRecord::Base.configurations
end

# db:load_config can be overriden manually
Rake::Task["db:seed"].enhance(["db:load_config"])
Rake::Task["db:load_config"].clear

# define Rails' tasks as no-op
Rake::Task.define_task("db:environment")
Rake::Task["db:test:deprecated"].clear if Rake::Task.task_defined?("db:test:deprecated")


lib/sinatra/activerecord/rake/activerecord_6.rb
require_relative "helpers"

seed_loader = Class.new do
  def load_seed
    load "#{ActiveRecord::Tasks::DatabaseTasks.db_dir}/seeds.rb"
  end
end

ActiveRecord::Tasks::DatabaseTasks.tap do |config|
  config.root                   = Rake.application.original_dir
  config.env                    = ENV["APP_ENV"] || ENV["RACK_ENV"] || "development"
  config.db_dir                 = "db"
  config.migrations_paths       = ["db/migrate"]
  config.fixtures_path          = "test/fixtures"
  config.seed_loader            = seed_loader.new
  config.database_configuration = ActiveRecord::Base.configurations
end

# db:load_config can be overriden manually
Rake::Task["db:seed"].enhance(["db:load_config"])
Rake::Task["db:load_config"].clear

# define Rails' tasks as no-op
Rake::Task.define_task("db:environment")
Rake::Task["db:test:deprecated"].clear if Rake::Task.task_defined?("db:test:deprecated")


lib/sinatra/activerecord.rb
require 'sinatra/base'
require 'active_record'
require 'active_support/core_ext/hash/keys'

require 'logger'
require 'pathname'
require 'yaml'
require 'erb'

require 'active_record/database_configurations/connection_url_resolver' if Gem.loaded_specs["activerecord"].version >= Gem::Version.create('6.1')

module Sinatra
  module ActiveRecordHelper
    def database
      settings.database
    end
  end

  module ActiveRecordExtension
    def self.registered(app)
      if ENV['DATABASE_URL'] && File.exist?("#{Dir.pwd}/config/database.yml")
        path = "#{Dir.pwd}/config/database.yml"
        url = ENV['DATABASE_URL']
        source = ERB.new(File.read(path)).result
        file_spec = YAML.respond_to?(:unsafe_load) ? YAML.unsafe_load(source) : YAML.load(source)
        file_spec ||= {}

        # ActiveRecord 6.1+ has moved the connection url resolver to another module
        if Gem.loaded_specs["activerecord"].version >= Gem::Version.create('6.1')
          url_spec = ActiveRecord::DatabaseConfigurations::ConnectionUrlResolver.new(url).to_hash
        else
          url_spec = ActiveRecord::ConnectionAdapters::ConnectionSpecification::ConnectionUrlResolver.new(url).to_hash
        end
        
        # if the configuration concerns only one database, and url_spec exist, url_spec will override the same key
        # if the configuration has multiple databases (Rails 6.0+ feature), url_spec is discarded
        # Following Active Record config convention
        # https://github.com/rails/rails/blob/main/activerecord/lib/active_record/database_configurations.rb#L169
        final_spec = file_spec.keys.map do |env|
          config = file_spec[env]
          if config.is_a?(Hash) && config.all? { |_k, v| v.is_a?(Hash) }
            [env, config]
          else
            [env, config.merge(url_spec)]
          end
        end.to_h

        app.set :database, final_spec
      elsif ENV['DATABASE_URL']
        app.set :database, ENV['DATABASE_URL']
      elsif File.exist?("#{Dir.pwd}/config/database.yml")
        app.set :database_file, "#{Dir.pwd}/config/database.yml"
      end

      unless defined?(Rake) || [:test, :production].include?(app.settings.environment)
        ActiveRecord::Base.logger = Logger.new(STDOUT)
      end

      app.helpers ActiveRecordHelper

      # re-connect if database connection dropped (Rails 3 only)
      app.before { ActiveRecord::Base.verify_active_connections! if ActiveRecord::Base.respond_to?(:verify_active_connections!) }
      app.after { ActiveRecord::Base.clear_active_connections! }
    end

    def database_file=(path)
      path = File.join(root, path) if Pathname(path).relative? and root
      source = ERB.new(File.read(path)).result
      spec = YAML.respond_to?(:unsafe_load) ? YAML.unsafe_load(source) : YAML.load(source)
      spec ||= {}
      set :database, spec
    end

    def database=(spec)
      if spec.is_a?(Hash) and spec.symbolize_keys[environment.to_sym]
        ActiveRecord::Base.configurations = spec.stringify_keys
        ActiveRecord::Base.establish_connection(environment.to_sym)
      elsif spec.is_a?(Hash)     
        ActiveRecord::Base.configurations = {
          environment.to_s => spec.stringify_keys
        }

        ActiveRecord::Base.establish_connection(spec.stringify_keys)
      else
        if Gem.loaded_specs["activerecord"].version >= Gem::Version.create('6.0')
          ActiveRecord::Base.configurations ||= ActiveRecord::DatabaseConfigurations.new({}).resolve(spec)
        else
          ActiveRecord::Base.configurations ||= {}
          ActiveRecord::Base.configurations[environment.to_s] = ActiveRecord::ConnectionAdapters::ConnectionSpecification::ConnectionUrlResolver.new(spec).to_hash
        end

        ActiveRecord::Base.establish_connection(spec)
      end
    end

    def database
      ActiveRecord::Base
    end
  end

  register ActiveRecordExtension
end
